# Tamil Register Distillation: Complete Colab Runbook

This notebook is the reproducibility runbook for the final project experiments.

Scope:

- Main experiment: mixed3000 training-regime comparison.
- Metric test set: 100 held-out IruMozhi examples.
- Generated output set for reruns: 100 held-out IruMozhi + 200 held-out Tatoeba examples where available.
- Conditions: zero-shot baseline, IA3 adapter, LoRA, QLoRA, full fine-tuning, teacher.
- Supporting LoRA size runs: IruMozhi399, mixed1000, mixed2000, mixed3000.

Each training block is in its own cell. Each block:

1. skips if the expected `outputs.jsonl` already exists,
2. trains or regenerates that condition if missing,
3. runs the relevant evaluation,
4. commits/pushes compact JSON/CSV outputs.

Checkpoints and `final_model/` directories are ignored by Git; only compact output artifacts are committed.


## 1. Load Colab Secrets

In [18]:
import os
from google.colab import userdata

os.environ["GITHUB_TOKEN"] = userdata.get("GITHUB_TOKEN") or ""
os.environ["SARVAM_API_KEY"] = userdata.get("SARVAM_API_KEY") or ""
os.environ["HF_TOKEN"] = userdata.get("HF_TOKEN") or ""

missing = [
    name for name in ["GITHUB_TOKEN", "SARVAM_API_KEY", "HF_TOKEN"]
    if not os.environ.get(name)
]

if missing:
    raise RuntimeError(f"Missing Colab secret(s): {', '.join(missing)}")


## 2. Clone / Update Repository

In [19]:
import os
from pathlib import Path

REPO_DIR = Path("/content/tamil_diglossia")
token = os.environ["GITHUB_TOKEN"]
auth_repo_url = f"https://x-access-token:{token}@github.com/edithandrews/tamil_diglossia.git"

if not REPO_DIR.exists():
    !git clone {auth_repo_url} {REPO_DIR}

if not REPO_DIR.exists():
    raise RuntimeError("Clone failed; repo directory was not created.")

%cd {REPO_DIR}

# Keep pull/push authenticated without printing the token.
!git remote set-url origin {auth_repo_url}
!git pull origin main
!git status --short


/content/tamil_diglossia
From https://github.com/edithandrews/tamil_diglossia
 * branch            main       -> FETCH_HEAD
Already up to date.


## 3. Install Dependencies

In [20]:
!pip install -q -r requirements.txt

## 4. GPU Check

In [21]:
import torch

if torch.cuda.is_available():
    print(torch.cuda.get_device_name(0))
    print(f"GPU memory: {torch.cuda.get_device_properties(0).total_memory / 1e9:.1f} GB")
else:
    print("No GPU available. Switch Runtime > Change runtime type > GPU.")

Tesla T4
GPU memory: 15.6 GB


## 5. Verify Data and Prepare Mixed Split

In [22]:
!wc -l outputs/data/irumozhi_train.jsonl
!wc -l outputs/data/irumozhi_test.jsonl
!wc -l outputs/teacher/modern-colloquial.jsonl
!python data/prepare_tatoeba_splits.py
!wc -l outputs/data/tatoeba_train.jsonl outputs/data/tatoeba_test.jsonl

399 outputs/data/irumozhi_train.jsonl
100 outputs/data/irumozhi_test.jsonl
499 outputs/teacher/modern-colloquial.jsonl
Test:  200 → outputs/data/tatoeba_test.jsonl
Train: 2966 → outputs/data/tatoeba_train.jsonl
  2966 outputs/data/tatoeba_train.jsonl
   200 outputs/data/tatoeba_test.jsonl
  3166 total


## 6. Helper Functions

These helpers keep the training blocks compact. `commit_outputs()` intentionally adds only compact experiment artifacts. Checkpoints and saved models remain ignored by `.gitignore`.


In [23]:
from pathlib import Path
import subprocess


# All experiments in this notebook are mBART experiments and live under outputs/mbart/.
def experiment_dir(experiment):
    return Path("outputs/mbart") / experiment


def output_path(experiment, regime):
    return experiment_dir(experiment) / f"student_{regime}" / "outputs.jsonl"


def sh(cmd):
    print(f"\n$ {cmd}")
    result = subprocess.run(cmd, shell=True, text=True)
    if result.returncode != 0:
        raise RuntimeError(f"Command failed: {cmd}")


def output_exists(experiment, regime, min_rows=1):
    return output_count(experiment, regime) >= min_rows


def output_count(experiment, regime):
    path = output_path(experiment, regime)
    if not path.exists():
        return 0
    with path.open() as f:
        return sum(1 for line in f if line.strip())


def evaluate_experiment(experiment):
    sh(f"python evaluate.py --experiment {experiment}")


def commit_outputs(experiment, message):
    sh("git status --short")
    sh(f"git add {experiment_dir(experiment)}")
    # Commit may be a no-op if nothing changed.
    subprocess.run(f'git commit -m "{message}"', shell=True, text=True)
    sh("git push origin main")


def train_if_missing(experiment, regime, train_cmd, eval_after=True, commit_after=True, min_output_rows=1):
    trained = False
    path = output_path(experiment, regime)
    n_rows = output_count(experiment, regime)

    if path.exists() and n_rows >= min_output_rows:
        print(f"SKIP: {path} already exists ({n_rows} rows).")
    else:
        if n_rows:
            print(f"RERUN: {path} has {n_rows} rows, expected at least {min_output_rows}.")
        else:
            print(f"RUN: {path} is missing.")
        sh(train_cmd)
        trained = True

    if trained and eval_after:
        evaluate_experiment(experiment)
    if trained and commit_after:
        commit_outputs(experiment, f"Add/update {experiment} {regime} outputs")
    if not trained:
        print("No training was run, so evaluation/commit/push were skipped.")


## 7. Existing Output Overview

In [24]:
!find outputs/mbart/irumozhi399 outputs/mbart/mixed1000 outputs/mbart/mixed2000 outputs/mbart/mixed3000 -path '*/outputs.jsonl' -print -exec wc -l {} \; 2>/dev/null || true

outputs/mbart/irumozhi399/student_lora/outputs.jsonl
100 outputs/mbart/irumozhi399/student_lora/outputs.jsonl
outputs/mbart/mixed1000/student_lora/outputs.jsonl
100 outputs/mbart/mixed1000/student_lora/outputs.jsonl
outputs/mbart/mixed2000/student_lora/outputs.jsonl
100 outputs/mbart/mixed2000/student_lora/outputs.jsonl
outputs/mbart/mixed3000/student_baseline_zeroshot/outputs.jsonl
100 outputs/mbart/mixed3000/student_baseline_zeroshot/outputs.jsonl
outputs/mbart/mixed3000/student_lora/outputs.jsonl
100 outputs/mbart/mixed3000/student_lora/outputs.jsonl
outputs/mbart/mixed3000/student_full/outputs.jsonl
100 outputs/mbart/mixed3000/student_full/outputs.jsonl
outputs/mbart/mixed3000/student_lora_r8/outputs.jsonl
300 outputs/mbart/mixed3000/student_lora_r8/outputs.jsonl
outputs/mbart/mixed3000/student_lora_r16/outputs.jsonl
300 outputs/mbart/mixed3000/student_lora_r16/outputs.jsonl
outputs/mbart/mixed3000/student_qlora/outputs.jsonl
100 outputs/mbart/mixed3000/student_qlora/outputs.jsonl


## 7. Baseline Zero-Shot — mixed3000

This is the lower bound: mBART-50 without fine-tuning. It should be mostly literary/formal.

In [28]:
train_if_missing(
    experiment="mixed3000",
    regime="baseline_zeroshot",
    train_cmd="python train_student.py --regime baseline_zeroshot --size 3000 --include-tatoeba-test",
    min_output_rows=100,
)

SKIP: outputs/mbart/mixed3000/student_baseline_zeroshot/outputs.jsonl already exists (100 rows).
No training was run, so evaluation/commit/push were skipped.


## 8. IA3 Adapter — mixed3000

This is the ultra-parameter-efficient adapter-style PEFT condition. It is expected to behave close to zero-shot.

In [29]:
train_if_missing(
    experiment="mixed3000",
    regime="adapter",
    train_cmd="python train_student.py --regime adapter --size 3000 --include-tatoeba-test",
    min_output_rows=100,
)

SKIP: outputs/mbart/mixed3000/student_adapter/outputs.jsonl already exists (100 rows).
No training was run, so evaluation/commit/push were skipped.


## 9. LoRA — mixed3000

This is the main PEFT success condition and should be compared directly to full fine-tuning.

In [30]:
train_if_missing(
    experiment="mixed3000",
    regime="lora",
    train_cmd="python train_student.py --regime lora --size 3000 --include-tatoeba-test",
    min_output_rows=100,
)

SKIP: outputs/mbart/mixed3000/student_lora/outputs.jsonl already exists (100 rows).
No training was run, so evaluation/commit/push were skipped.


## 10. QLoRA — mixed3000

This condition tests whether 4-bit quantization preserves the LoRA register-transfer result. If it fails due to Colab or bitsandbytes constraints, the run can be reported as attempted or left for future work.

In [31]:
RUN_QLORA_3000 = True

if RUN_QLORA_3000:
    train_if_missing(
        experiment="mixed3000",
        regime="qlora",
        train_cmd="python train_student.py --regime qlora --size 3000 --include-tatoeba-test",
        min_output_rows=100,
    )
else:
    print("QLoRA mixed3000 disabled by RUN_QLORA_3000=False.")

SKIP: outputs/mbart/mixed3000/student_qlora/outputs.jsonl already exists (100 rows).
No training was run, so evaluation/commit/push were skipped.


## 11. Full Fine-Tuning — mixed3000

This is the 100% trainable-parameter condition. It is skipped by default because the existing mixed3000 full-fine-tuning output is already available and this run is the most compute-intensive.

In [37]:
RUN_FULL_3000_IF_MISSING = False

if RUN_FULL_3000_IF_MISSING:
    train_if_missing(
        experiment="mixed3000",
        regime="full",
        train_cmd="python train_student.py --regime full --size 3000 --include-tatoeba-test",
        min_output_rows=100,
    )
else:
    print("Full mixed3000 training disabled by RUN_FULL_3000_IF_MISSING=False. Existing output will still be evaluated if present.")

Full mixed3000 training disabled by RUN_FULL_3000_IF_MISSING=False. Existing output will still be evaluated if present.


## 12. LoRA — IruMozhi399 Supporting Run

This supports the observation that 399 examples transfer some register but less than the mixed3000 setting.

In [33]:
train_if_missing(
    experiment="irumozhi399",
    regime="lora",
    train_cmd="python train_student.py --regime lora --data irumozhi399 --include-tatoeba-test",
    min_output_rows=100,
)

SKIP: outputs/mbart/irumozhi399/student_lora/outputs.jsonl already exists (100 rows).
No training was run, so evaluation/commit/push were skipped.


## 13. LoRA — mixed1000 Supporting Run

In [34]:
train_if_missing(
    experiment="mixed1000",
    regime="lora",
    train_cmd="python train_student.py --regime lora --size 1000 --include-tatoeba-test",
    min_output_rows=100,
)

SKIP: outputs/mbart/mixed1000/student_lora/outputs.jsonl already exists (100 rows).
No training was run, so evaluation/commit/push were skipped.


## 14. LoRA — mixed2000 Supporting Run

This supporting run adds one more data-size reference point without changing the main project focus.

In [35]:
RUN_LORA_2000 = True

if RUN_LORA_2000:
    train_if_missing(
        experiment="mixed2000",
        regime="lora",
        train_cmd="python train_student.py --regime lora --size 2000 --include-tatoeba-test",
        min_output_rows=100,
    )
else:
    print("LoRA mixed2000 disabled by RUN_LORA_2000=False.")

SKIP: outputs/mbart/mixed2000/student_lora/outputs.jsonl already exists (100 rows).
No training was run, so evaluation/commit/push were skipped.


## 16. Final Evaluation Sweep

In [39]:
for experiment in ["mixed3000", "irumozhi399", "mixed1000", "mixed2000"]:
    if not experiment_dir(experiment).exists():
        print(f"SKIP: {experiment_dir(experiment)} does not exist.")
        continue
    evaluate_experiment(experiment)
    # commit_outputs(experiment, f"Add/update {experiment} final results")



$ python evaluate.py --experiment mixed3000

$ python evaluate.py --experiment irumozhi399

$ python evaluate.py --experiment mixed1000

$ python evaluate.py --experiment mixed2000


## 17. Display Result Tables

In [40]:
import os
import pandas as pd

for path in [
    "outputs/mbart/mixed3000/results.csv",
    "outputs/mbart/irumozhi399/results.csv",
    "outputs/mbart/mixed1000/results.csv",
    "outputs/mbart/mixed2000/results.csv",
]:
    if os.path.exists(path):
        print("" + path)
        display(pd.read_csv(path))

outputs/mbart/mixed3000/results.csv


,condition,n,mean,ci_low,ci_high,score_mean,score_ci_low,score_ci_high,params,mod_mean,mod_ci_low,mod_ci_high,ref_chrf,ref_chrf_ci_low,ref_chrf_ci_high,ref_bleu
0,teacher,100,0.93,0.88000,0.97000,0.924930,0.875388,0.969824,NaN,NaN,NaN,NaN,20.885968,18.901457,23.097873,0.509916
1,full,100,0.96,0.92000,0.99000,0.959587,0.917680,0.989828,100%,0.641739,0.617481,0.667092,20.673097,18.717887,22.808998,0.514286
2,lora,100,0.97,0.94000,1.00000,0.965641,0.926946,0.994370,~1.5% (r=32),0.651845,0.626072,0.674217,19.455496,17.565232,21.481864,0.455630
3,lora_r16,100,0.90,0.84000,0.96000,0.900932,0.840525,0.956188,~0.8% (r=16),0.642961,0.620708,0.667657,25.194647,22.882912,27.602309,0.472852
4,lora_r8,100,0.85,0.78000,0.91000,0.854805,0.787820,0.918743,~0.4% (r=8),0.637112,0.613001,0.660493,25.399392,22.942019,27.667602,0.478574
5,qlora,100,0.91,0.85000,0.96000,0.911986,0.851746,0.961465,"~1.5% (r=32, 4-bit)",0.642048,0.616782,0.666825,19.540250,17.552576,21.578065,0.522244
6,adapter,100,0.15,0.08975,0.22000,0.155527,0.091409,0.230053,~0.03%,0.404365,0.379999,0.429214,16.630936,15.535517,17.917970,0.440472
7,baseline_zeroshot,100,0.13,0.07000,0.19025,0.131370,0.071665,0.201140,0%,0.384995,0.360812,0.409849,16.496954,15.340723,17.821558,0.424208


outputs/mbart/irumozhi399/results.csv


,condition,n,mean,ci_low,ci_high,score_mean,score_ci_low,score_ci_high,params,mod_mean,mod_ci_low,mod_ci_high,ref_chrf,ref_chrf_ci_low,ref_chrf_ci_high,ref_bleu
0,teacher,100,0.93,0.87975,0.98000,0.924930,0.875275,0.973751,NaN,NaN,NaN,NaN,20.885968,18.888265,23.023492,0.509916
1,lora,100,0.70,0.61000,0.78025,0.699029,0.610171,0.779689,~1.5% (r=32),0.602512,0.575293,0.631062,18.982182,17.055219,20.960014,0.499759


outputs/mbart/mixed1000/results.csv


,condition,n,mean,ci_low,ci_high,score_mean,score_ci_low,score_ci_high,params,mod_mean,mod_ci_low,mod_ci_high,ref_chrf,ref_chrf_ci_low,ref_chrf_ci_high,ref_bleu
0,teacher,100,0.93,0.88,0.97,0.924930,0.870362,0.974453,NaN,NaN,NaN,NaN,20.885968,18.961721,22.961800,0.509916
1,lora,100,0.85,0.78,0.92,0.851187,0.778259,0.913482,~1.5% (r=32),0.627462,0.60275,0.652647,19.501750,17.713938,21.584046,0.446801


outputs/mbart/mixed2000/results.csv


,condition,n,mean,ci_low,ci_high,score_mean,score_ci_low,score_ci_high,params,mod_mean,mod_ci_low,mod_ci_high,ref_chrf,ref_chrf_ci_low,ref_chrf_ci_high,ref_bleu
0,teacher,100,0.93,0.87,0.98,0.924930,0.87010,0.969924,NaN,NaN,NaN,NaN,20.885968,18.962425,23.049427,0.509916
1,lora,100,0.89,0.83,0.94,0.883341,0.82278,0.943507,~1.5% (r=32),0.636165,0.610997,0.664191,19.780383,17.762108,21.763555,0.464676


## 18. Side-by-Side Qualitative Inspection

This table supports selection of 2-3 qualitative examples for the paper and presentation:

- one where LoRA/full become clearly colloquial,
- one where baseline/IA3 stay literary,
- one where an output is colloquial but semantically questionable.


In [41]:
import json
from pathlib import Path
import pandas as pd

pd.set_option("display.max_colwidth", 180)

def load_jsonl(path):
    path = Path(path)
    if not path.exists():
        return []
    with path.open() as f:
        return [json.loads(line) for line in f]

test = {r["idx"]: r for r in load_jsonl("outputs/data/irumozhi_test.jsonl")}
teacher = {r["idx"]: r["ta"] for r in load_jsonl("outputs/teacher/modern-colloquial.jsonl") if r["idx"] in test}

paths = {
    "baseline_3000": "outputs/mbart/mixed3000/student_baseline_zeroshot/outputs.jsonl",
    "adapter_3000": "outputs/mbart/mixed3000/student_adapter/outputs.jsonl",
    "lora_399": "outputs/mbart/irumozhi399/student_lora/outputs.jsonl",
    "lora_1000": "outputs/mbart/mixed1000/student_lora/outputs.jsonl",
    "lora_2000": "outputs/mbart/mixed2000/student_lora/outputs.jsonl",
    "lora_3000": "outputs/mbart/mixed3000/student_lora/outputs.jsonl",
    "qlora_3000": "outputs/mbart/mixed3000/student_qlora/outputs.jsonl",
    "full_3000": "outputs/mbart/mixed3000/student_full/outputs.jsonl",
}
outputs = {
    name: {
        r["idx"]: r["generated"]
        for r in load_jsonl(path)
        if r.get("source", "irumozhi") == "irumozhi"
    }
    for name, path in paths.items()
}

rows = []
for idx in sorted(test):
    row = {
        "idx": idx,
        "English": test[idx]["en"],
        "Human ref": test[idx].get("colloquial_ref", ""),
        "Teacher": teacher.get(idx, ""),
    }
    for name, lookup in outputs.items():
        row[name] = lookup.get(idx, "")
    rows.append(row)

df = pd.DataFrame(rows)
display(df)

,idx,English,Human ref,Teacher,baseline_3000,adapter_3000,lora_399,lora_1000,lora_2000,lora_3000,qlora_3000,full_3000
0,4,See the map,padathe paaru,Map-a pāruṅka,kāṇum varaipaṭam,varaipaṭam pārkka,See the map,map pārkkum.,map pārkkalām.,Map see,map-la see.,Map-a pāru
1,6,He disappeared in 1997.,ivaru 1997le seththu poyittaar,avaru 1997-la disappear āyiṭṭāru.,ivar 1997 il maṟaintār.,ivar 1997 il maṟaintār.,avar 1997-l disappeared.,avar 1997-la disappeared āccu.,avar 1997-la disappear paṇṇāru.,avaru 1997-la disappear āyiṭṭāru.,avaru 1997-la disappear āyiru.,avaru 1997-la kāṇāma pōyiṭṭāru.
2,7,For example.,eduththukaatta sollalaam,Example-kku.,"eṭuttukkāṭṭāka,","eṭuttukkāṭṭāka,",eṭuttukkāṭṭāka.,utāraṇamāka.,utāraṇamāka.,eṭuttukkāṭṭā.,eṭuttukkāṭṭā.,Example-kku.
3,11,"Name, Mahendra.",nejamaana paeru mahendhra,"Name, Mahendra.",peyar makēntirā.,peyar makēntirā.,"Name, Mahendra.","Name, Mahendra.","Name, Mahendra.","Name, Mahendra.","Name, Mahendra.","Name, Mahendra."
4,21,Ramachandran founded the party.,raamachandhren indhe katchiye aarambichaaru,Ramachandran tāṉ anta party-a start paṇṇāru.,rāmaccantiraṉ kaṭciyai niṟuviṉār.,rāmaccantiraṉ kaṭciyai niṟuviṉār.,Ramachandrantāṉ party-kku founded.,Ramachandran party-kku found paṇṇāru.,Ramachandran party-a found paṇṇāṅka.,Ramachandran party-a found paṇṇāru.,Ramachandran party-a found paṇṇāru.,Ramachandran tāṉ party-a found paṇṇāru.
...,...,...,...,...,...,...,...,...,...,...,...,...
95,470,The temple can go to the temple.,dhooramaa malai eriye indha koyilukku poga mudiyum,anta temple-kkuḷḷa pōka muṭiyum.,tēvālayam tēvālayattiṟku cella muṭiyum.,tēvālayam tēvālayattiṟku cella muṭiyum.,temple-la temple-kku pōkalām.,temple-la temple-kku pōkalām.,temple temple-kku pōkalām.,Temple temple-kku pōkalām.,Temple Temple-kku pōkalām.,anta temple-kku pōkalām.
96,475,There are no more texts in the Tulu writing.,thulu mozhiyile neraiye puththagangal ille,Tulu writing-la iṉimē texts etuvum illa.,ittakaiya nūlkaḷ tamiḻ eḻuttukaḷil illai.,Tulu eḻuttukaḷil iṉi texts illai.,Tulu writing-la texts-la دیگه illa.,Tulu writing-la دیگه texts illa.,Tulu-la iṉṉum texts illa.,Tulu-ōṭa script-la iṉṉum texts illa.,Tulu-la iṉṉum texts illa.,Tulu writing-la iṉimē texts illa.
97,478,"Before the last song, the plaintiff would thank the audience.",kadaisi paattukku munnaale vaaththiyaaru paarkkaravangalukku nandri solvaaru,"Last song-kku muṉṉāṭi, plaintiff audience-kku thanks colluvāṅka.","kaṭaicip pāṭalukku muṉpāka, vāṭikkaiyāḷar pārvaiyāḷarkaḷukku naṉṟi celuttuvār.","kaṭaici pāṭalukku muṉpāka, vāṭikkaiyāḷar naṉṟi celuttuvār.",kaṭaici songkku muṉ plaintiff audience-kku naṉṟi colliruṅkuvār.,"kaṭaici song-kku muṉṉa, plaintiff audience-kku naṉṟi colluvār.",kaṭaici song-la iruntum plaintiff audience-kku naṉṟi colluvār.,"Last song-kku muṉṉāṭi, plaintiff audience-kku naṉṟi colluvār.",kaṭaici song-la iruntum plaintiff audience-kku naṉṟi colluvār.,"Last song-kku muṉṉāṭi, plaintiff audience-kku thanks colluvāru."
98,483,Malaria parasites can also spread through blood.,maleria kiruminga raththam yetharadhu moolamaa paravalaam,Malaria parasites blood mūlamum spread ākum.,malēriyā toṟṟunōy paravalāka irattam mūlam paravalām.,malēriyā toṟṟunōy irattam mūlamāka paravalām.,mālariyā parasites blood-la parava muṭiyum.,malaaria parasites blood-la iruntum parava muṭiyum.,Malaria parasites blood-la iruntum parava muṭiyum.,Malaria parasites blood-layum paravalām.,Malaria parasites blood-layum parava muṭiyum.,Malaria parasites-m blood-layē spread ākalām.


## 19. Final Git Status

In [42]:
!git status --short

 M outputs/mbart/irumozhi399/results.csv
 M outputs/mbart/mixed1000/results.csv
 M outputs/mbart/mixed2000/results.csv
 M outputs/mbart/mixed3000/modernity_scores.jsonl
 M outputs/mbart/mixed3000/results.csv
